Add anonymous analytics for tracking command usage

Piccirello opened the pull request:
This allows us to see what commands people are most frequently using. Analytics are enabled by default and are completely anonymous.

Data we collect:
- command name
- os
- architecture
- CLI version

To opt out: run `doppler analytics disable`.

Closes ENG-735.

Commit:
Add anonymous analytics for tracking command usage

This allows us to see what commands people are most frequently using. Analytics are enabled by default and are completely anonymous.

Data we collect:
- command name
- os
- architecture
- CLI version

To opt out: run `doppler analytics disable`.

Kunamatata commented on pkg/cmd/root.go:
```diff
@@ -42,6 +43,10 @@ var rootCmd = &cobra.Command{
 		configuration.LoadConfig()
 
+		if configuration.IsAnalyticsEnabled() {
+			controllers.CaptureCommand(strings.ReplaceAll(cmd.CommandPath(), " ", "."))
```
Should it be the responsibility of `controllers.CaptureCommand(...)` to format the `cmd.CommandPath()` via `strings.ReplaceAll(cmd.CommandPath(), " ", ".")` instead of the caller doing it?

Kunamatata commented on pkg/http/analytics.go:
```diff
@@ -0,0 +1,39 @@
+)
+
+func CaptureCommand(command string) ([]byte, Error) {
+	postBody := map[string]interface{}{"command": command}
```
Do we want to have a struct for these instead of using `map[string]interface{}`? 
`map[string]interface{}` is comparable to typing with `any` in typescript, we lose the ability to rely on types for our objects.

```go
type CaptureCommand struct {
	Command string `json:"command"`
}
```

Piccirello replied:
Since we immediately marshal to json on the next line, I'm not sure that the struct gets us anything.

Kunamatata commented on pkg/cmd/analytics.go:
```diff
@@ -0,0 +1,88 @@
+	Args:  cobra.NoArgs,
+	Run: func(cmd *cobra.Command, args []string) {
+		if utils.OutputJSON {
+			printer.JSON(map[string]bool{"enabled": configuration.IsAnalyticsEnabled()})
```
Same comment about leveraging a struct instead of map[string]bool:
```go
type AnalyticsStatus struct {
	Enabled bool `json:"enabled"`
}
```

Piccirello replied:
Here too - we can add a struct, but I don't think we'd really get any benefit.

nmanoogian commented on pkg/controllers/analytics.go:
```diff
@@ -0,0 +1,34 @@
+
+func CaptureCommand(command string) {
+	// wait for analytics command to complete before exiting
+	ch := make(chan bool, 1)
```
What's happening with this channel here? Won't this function just block until `CaptureCommand` finishes?

Kunamatata replied:
https://pkg.go.dev/sync#WaitGroup Might be a good feature to leverage. 
`wg.Wait()` could happen as the last line of the `Execute()` function in `root.go` and allow wg to be accessed by goroutines in our code.
The first line of the `CaptureCommand` function could be `wg.Add()` and the first line inside the goroutine function `defer wg.Done()`

We can then get rid of the channel. This will ensure that the `Execute()` function does not end (until the wait group counter is done to 0) which would end the main thread and end all goroutines.

EDIT:
I just thought about this but we can potentially also have a wait group at the callsite of `CaptureCommand` (`root.go`) and pass the wait group inside the CaptureCommand.

Piccirello replied:
This indeed was not working as I had intended. I've switched to using a WaitGroup for this.

nmanoogian commented on pkg/http/analytics.go:
```diff
@@ -0,0 +1,39 @@
+	postBody := map[string]interface{}{"command": command}
+	body, err := json.Marshal(postBody)
+	if err != nil {
+		return nil, Error{Err: err, Message: "Unable to marshall command"}
```
nit: `marshal` (just one L)

nmanoogian commented on pkg/http/analytics.go:
```diff
@@ -0,0 +1,39 @@
+
+	_, _, resp, err := PostRequest("https://cli.doppler.com", true, map[string]string{"Content-Type": "application/json"}, "/v1/analytics", nil, body)
+	if err != nil {
+		return nil, Error{Err: err, Message: "Unable to fetch changelog"}
```
nit: copy-pasta

Commit:
chore: add analytics to version upgrade flow

This will let us see how frequently users are accepting the upgrade prompt.

nmanoogian commented on pkg/controllers/analytics.go:
```diff
@@ -30,3 +30,9 @@ func CaptureCommand(wg *sync.WaitGroup, command string) {
+func CaptureEvent(wg *sync.WaitGroup, event string) {
+	defer wg.Done()
+
+	http.CaptureEvent(event)
```
nit: To be very explicit that we're ignoring these returned values (and swallowing errors), it might be nice to assign the returned values to underscore:

```diff
diff --git a/pkg/controllers/analytics.go b/pkg/controllers/analytics.go
index 6d07189..4637176 100644
--- a/pkg/controllers/analytics.go
+++ b/pkg/controllers/analytics.go
@@ -28,11 +28,11 @@ func CaptureCommand(wg *sync.WaitGroup, command string) {
 	defer wg.Done()
 
 	command = strings.ReplaceAll(command, " ", ".")
-	http.CaptureCommand(command)
+	_, _ = http.CaptureCommand(command)
 }
 
 func CaptureEvent(wg *sync.WaitGroup, event string) {
 	defer wg.Done()
 
-	http.CaptureEvent(event)
+	_, _ = http.CaptureEvent(event)
 }


```

Piccirello replied:
Ended up debug logging the error message

nmanoogian commented on pkg/http/analytics.go:
```diff
@@ -37,3 +37,19 @@ func CaptureCommand(command string) ([]byte, Error) {
+	postBody := map[string]interface{}{"event": event}
+	body, err := json.Marshal(postBody)
+	if err != nil {
+		return nil, Error{Err: err, Message: "Unable to marshal command"}
```
nit: "event"

nmanoogian commented on pkg/cmd/root.go:
```diff
@@ -101,6 +101,9 @@ func checkVersion(command string) {
 	}
 
+	wg.Add(1)
+	go controllers.CaptureEvent(wg, "VersionCheck")
```
Should these be wrapped with `EnableAnalytics` checks?

Piccirello replied:
Good catch. I'm going to move the `IsAnalyticsEnabled` check into the analytics controller to avoid this issue in the future.

dp-franklin commented on pkg/http/analytics.go:
```diff
@@ -0,0 +1,55 @@
+
+func CaptureCommand(command string) ([]byte, Error) {
+	postBody := map[string]interface{}{"command": command}
+	body, err := json.Marshal(postBody)
```
Is there anything else we might want to capture that isn't intrusive?
Ideas:
- True/false: is `api-host` set to `https://api.doppler.com`?
- True/false: is `token` set?
- True/false: is the `scope` set to `/`?

Piccirello replied:
The scope one is particularly interesting. I have wondered how frequently users are specifying a scope rather than using the default global scope (specifically when logging in and when running setup). I'll create a separate ticket for this.

Piccirello replied:
Created ENG-5666

dp-franklin commented on tests/e2e/analytics.sh:
```diff
@@ -0,0 +1,59 @@
+beforeEach
+
+# analytics defaults to enabled after clearing config
+"$DOPPLER_BINARY" configure reset --configuration=./temp-config --yes
```
Nit: I think it would be better to test the `doppler configure reset` command *without* first deleting the existing temp-config file in `beforeEach`

Piccirello replied:
Good call, I'll add an additional test to verify `reset` works as expected when there's an existing config with analytics disabled.

dp-franklin commented on pkg/cmd/root.go:
```diff
@@ -31,6 +32,7 @@ import (
 	"gopkg.in/gookit/color.v1"
 )
 
+var wg *sync.WaitGroup
```
Shouldn't `wg.Wait()` be called at some point?

Piccirello replied:
We call `Wait()` at the bottom of this file in `Execute()`

dp-franklin replied:
Ah, I see now. Nevermind.

Commit:
chore: move keyring impl to configuration package

This prevent a circular import between the configuration and controllers packages.

Commit:
chore: move goreleaser check to separate Action workflow

This allows us to set the check as optional as deprecation notices shouldn't prevent merging.

dp-franklin commented on pkg/http/analytics.go:
```diff
@@ -0,0 +1,55 @@
+
+	utils.LogDebug(fmt.Sprintf("Sending anonymous analytics payload: '%s'", body))
+
+	_, _, resp, err := PostRequest("https://cli.doppler.com", true, map[string]string{"Content-Type": "application/json"}, "/v1/analytics", nil, body)
```
Question: is 10 seconds a reasonable amount timeout for sending analytics? I think it is but I'd like to get a second opinion. The concern is that this could prevent the CLI from exiting in a timely manner for an otherwise simple command.

Piccirello replied:
10 seconds is likely longer than we need, especially because we don't actually care about the response. This will require a bit of rework given the current structure of our http package, so I'd opt for improving this in a follow up PR.
